File: day-5/ruby/lib/glyphs/letters.rb
#!/usr/bin/env ruby
require_relative '../glyph_dsl'

GlyphDSL.define_glyph ' ' do
  # pass :)
end

# Define all letter glyphs using the DSL with normalized coordinates (0-100)
GlyphDSL.define_glyph :g do
  # G is a circle with a line extending from the bottom right
  circle 50, 50, 50
  arc 50, 10, 50, 240, 390
end

GlyphDSL.define_glyph :e do
  # E is a 3/4 circle with a horizontal line
  arc 50, 50, 50, 0, 270
  line 50, 50, 100, 50
end

GlyphDSL.define_glyph :n do
  # N is a vertical line, half circle, and another vertical line
  vertical_line 0, 0, 100
  arc 50, 50, 50, 0, 180
  vertical_line 100, 0, 50
end

GlyphDSL.define_glyph :u do
  # U is a vertical line, half circle at bottom, and another vertical line
  vertical_line 0, 100, 50
  arc 50, 50, 50, 180, 360
  vertical_line 100, 100, 50
end

GlyphDSL.define_glyph :r do
  # R is a vertical line with a quarter circle at top right
  vertical_line 0, 0, 100
  arc 0, 50, 50, 0, 90
  # Make this letter narrower than default
  restrict_width 10
end

GlyphDSL.define_glyph :a do
  # A is a circle with a vertical line on the right
  circle 50, 50, 50
  vertical_line 100, 0, 100
end

GlyphDSL.define_glyph :y do
  # Y is a V shape with a stem
  line 0, 100, 50, 50
  line 50, 50, 100, 100
  vertical_line 50, 50, 0
end

# Additional letters
GlyphDSL.define_glyph :b do
  vertical_line 0, 0, 100
  arc 50, 25, 50, 270, 90
  arc 50, 75, 50, 270, 90
end

GlyphDSL.define_glyph :c do
  arc 50, 50, 50, 45, 315
  # Make this letter narrower than default
  restrict_width 70
end

GlyphDSL.define_glyph :d do
  vertical_line 100, 0, 100
  arc 50, 50, 50, 90, 270
end

GlyphDSL.define_glyph :f do
  vertical_line 25, 0, 100
  horizontal_line 25, 75, 100
  horizontal_line 25, 75, 50
  # Make this letter narrower than default
  restrict_width 75
end

GlyphDSL.define_glyph :h do
  vertical_line 0, 0, 100
  vertical_line 100, 0, 100
  horizontal_line 0, 100, 50
end

GlyphDSL.define_glyph :i do
  vertical_line 50, 0, 100
  circle 50, 100, 5
  # Make this letter much narrower
  restrict_width 30
end

GlyphDSL.define_glyph :j do
  vertical_line 75, 0, 75
  arc 50, 25, 25, 0, 180
  circle 75, 100, 5
  # Make this letter narrower
  restrict_width 75
end

GlyphDSL.define_glyph :k do
  vertical_line 0, 0, 100
  line 0, 50, 100, 100
  line 0, 50, 100, 0
  # Make this letter narrower
  restrict_width 80
end

GlyphDSL.define_glyph :l do
  vertical_line 50, 0, 100
  # Make this letter much narrower
  restrict_width 30
end

GlyphDSL.define_glyph :m do
  vertical_line 0, 0, 100
  vertical_line 50, 0, 100
  vertical_line 100, 0, 100
  arc 25, 100, 25, 180, 0
  arc 75, 100, 25, 180, 0
end

GlyphDSL.define_glyph :o do
  circle 50, 50, 50
  # Make this letter slightly narrower
  restrict_width 90
end

GlyphDSL.define_glyph :p do
  vertical_line 0, 0, 100
  arc 50, 75, 50, 270, 90
  # Make this letter narrower
  restrict_width 80
end

GlyphDSL.define_glyph :q do
  circle 50, 50, 50
  line 70, 30, 100, 0
end

GlyphDSL.define_glyph :s do
  arc 50, 75, 25, 0, 270
  arc 50, 25, 25, 180, 90
  # Make this letter narrower
  restrict_width 70
end

GlyphDSL.define_glyph :t do
  vertical_line 50, 0, 100
  horizontal_line 25, 75, 100
  # Make this letter narrower
  restrict_width 70
end

GlyphDSL.define_glyph :v do
  line 0, 100, 50, 0
  line 50, 0, 100, 100
end

GlyphDSL.define_glyph :w do
  line 0, 100, 25, 0
  line 25, 0, 50, 50
  line 50, 50, 75, 0
  line 75, 0, 100, 100
end

GlyphDSL.define_glyph :x do
  line 0, 0, 100, 100
  line 0, 100, 100, 0
end

GlyphDSL.define_glyph :z do
  horizontal_line 0, 100, 100
  line 100, 100, 0, 0
  horizontal_line 0, 100, 0
  # Make this letter narrower
  restrict_width 80
end

# Add basic numbers
GlyphDSL.define_glyph '0' do
  circle 50, 50, 50
  # Make this number narrower
  restrict_width 80
end

GlyphDSL.define_glyph '1' do
  vertical_line 50, 0, 100
  # Make this number much narrower
  restrict_width 30
end

GlyphDSL.define_glyph '2' do
  arc 50, 75, 50, 0, 270
  line 50, 75, 0, 0
  # Make this number narrower
  restrict_width 80
end

File: day-5/rust/src/letters.rs
use crate::types::{Pt, Stroke, Glyph};
use std::collections::HashMap;

macro_rules! glyphs {
    (
        $(
            $ch:literal {
                $($body:tt)*
            }
        )*
    ) => {{
        let mut glyphs = HashMap::new();
        $(
            let (strokes, width_restriction) = glyphs!(@parse_strokes $($body)*);
            let glyph = Glyph {
                ch: $ch,
                strokes,
                width_restriction,
            };
            glyphs.insert(glyph.ch, glyph);
        )*
        glyphs
    }};

    // top level recursive parsing
    (@parse_strokes $($tt:tt)*) => {{
        let mut strokes_vec: Vec<Stroke> = Vec::new();
        let mut width_restriction: Option<u8> = None;
        glyphs!(@strokes strokes_vec, width_restriction, $($tt)*);
        (strokes_vec, width_restriction)
    }};

    // Base case: no more strokes
    (@strokes $out:ident, $width:ident, ) => {};

    // lines grammar
    (@strokes $out:ident, $width:ident,
        line ($x1:expr, $y1:expr) -> ($x2:expr, $y2:expr)
        $($rest:tt)*
    ) => {{
        $out.push(Stroke::Line {
            from: Pt($x1, $y1),
            to: Pt($x2, $y2),
        });
        glyphs!(@strokes $out, $width, $($rest)*);
    }};

    // arcs grammar
    (@strokes $out:ident, $width:ident,
        arc ($x:expr, $y:expr) radius ($r:expr) from ($start:expr) -> ($end:expr)
        $($rest:tt)*
    ) => {{
        $out.push(Stroke::Arc {
            center: Pt($x, $y),
            radius: $r,
            start: $start,
            end: $end,
        });
        glyphs!(@strokes $out, $width, $($rest)*);
    }};

    // restrict_width grammar
    (@strokes $out:ident, $width:ident,
        restrict_width ($w:expr)
        $($rest:tt)*
    ) => {{
        $width = Some($w as u8);
        glyphs!(@strokes $out, $width, $($rest)*);
    }};
}

pub fn define_glyphs() -> HashMap<char, Glyph> {
    let glyphs = glyphs! {
        // Space
        ' ' {
            // Empty glyph
        }

        // Letter g - arc with a line extending from the bottom right
        'g' {
            arc (50, 50) radius (50) from (0) -> (360)
            arc (50, 10) radius (50) from (240) -> (390)
        }

        // Letter e - 3/4 circle with a horizontal line
        'e' {
            arc (50, 50) radius (50) from (0) -> (270)
            line (50, 50) -> (100, 50)
        }

        // Letter n - vertical line, half circle, and another vertical line
        'n' {
            line (0, 0) -> (0, 100)
            arc (50, 50) radius (50) from (0) -> (180)
            line (100, 0) -> (100, 50)
        }

        // Letter u - vertical line, half circle at bottom, and another vertical line
        'u' {
            line (0, 100) -> (0, 50)
            arc (50, 50) radius (50) from (180) -> (360)
            line (100, 100) -> (100, 50)
        }

        // Letter r - vertical line with a quarter circle at top right
        'r' {
            line (0, 0) -> (0, 100)
            arc (0, 50) radius (50) from (0) -> (90)
            restrict_width (10)
        }

        // Letter a - circle with a vertical line on the right
        'a' {
            arc (50, 50) radius (50) from (0) -> (360)
            line (100, 0) -> (100, 100)
        }

        // Letter y - V shape with a stem
        'y' {
            line (0, 100) -> (50, 50)
            line (50, 50) -> (100, 100)
            line (50, 50) -> (50, 0)
        }

        // Letter b
        'b' {
            line (0, 0) -> (0, 100)
            arc (50, 25) radius (50) from (270) -> (90)
            arc (50, 75) radius (50) from (270) -> (90)
        }

        // Letter c
        'c' {
            arc (50, 50) radius (50) from (45) -> (315)
            restrict_width (70)
        }

        // Letter d
        'd' {
            line (100, 0) -> (100, 100)
            arc (50, 50) radius (50) from (90) -> (270)
        }

        // Letter f
        'f' {
            line (25, 0) -> (25, 100)
            line (25, 100) -> (75, 100)
            line (25, 50) -> (75, 50)
            restrict_width (75)
        }

        // Letter h
        'h' {
            line (0, 0) -> (0, 100)
            line (100, 0) -> (100, 100)
            line (0, 50) -> (100, 50)
        }

        // Letter i
        'i' {
            line (50, 0) -> (50, 100)
            arc (50, 100) radius (5) from (0) -> (360)
            restrict_width (30)
        }

        // Letter j
        'j' {
            line (75, 0) -> (75, 75)
            arc (50, 25) radius (25) from (0) -> (180)
            arc (75, 100) radius (5) from (0) -> (360)
            restrict_width (75)
        }

        // Letter k
        'k' {
            line (0, 0) -> (0, 100)
            line (0, 50) -> (100, 100)
            line (0, 50) -> (100, 0)
            restrict_width (80)
        }

        // Letter l
        'l' {
            line (50, 0) -> (50, 100)
            restrict_width (30)
        }

        // Letter m
        'm' {
            line (0, 0) -> (0, 100)
            line (50, 0) -> (50, 100)
            line (100, 0) -> (100, 100)
            arc (25, 100) radius (25) from (180) -> (0)
            arc (75, 100) radius (25) from (180) -> (0)
        }

        // Letter o
        'o' {
            arc (50, 50) radius (50) from (0) -> (360)
            restrict_width (90)
        }

        // Letter p
        'p' {
            line (0, 0) -> (0, 100)
            arc (50, 75) radius (50) from (270) -> (90)
            restrict_width (80)
        }

        // Letter q
        'q' {
            arc (50, 50) radius (50) from (0) -> (360)
            line (70, 30) -> (100, 0)
        }

        // Letter s
        's' {
            arc (50, 75) radius (25) from (0) -> (270)
            arc (50, 25) radius (25) from (180) -> (90)
            restrict_width (70)
        }

        // Letter t
        't' {
            line (50, 0) -> (50, 100)
            line (25, 100) -> (75, 100)
            restrict_width (70)
        }

        // Letter v
        'v' {
            line (0, 100) -> (50, 0)
            line (50, 0) -> (100, 100)
        }

        // Letter w
        'w' {
            line (0, 100) -> (25, 0)
            line (25, 0) -> (50, 50)
            line (50, 50) -> (75, 0)
            line (75, 0) -> (100, 100)
        }

        // Letter x
        'x' {
            line (0, 0) -> (100, 100)
            line (0, 100) -> (100, 0)
        }

        // Letter z
        'z' {
            line (0, 100) -> (100, 100)
            line (100, 100) -> (0, 0)
            line (0, 0) -> (100, 0)
            restrict_width (80)
        }

        // Number 0
        '0' {
            arc (50, 50) radius (50) from (0) -> (360)
            restrict_width (80)
        }

        // Number 1
        '1' {
            line (50, 0) -> (50, 100)
            restrict_width (30)
        }

        // Number 2
        '2' {
            arc (50, 75) radius (50) from (0) -> (270)
            line (50, 75) -> (0, 0)
            restrict_width (80)
        }
    };

    glyphs
}